src/eduid/webapp/idp/tests/test_login.py

```diff
@@ -82,28 +83,129 @@ def test_login_pwauth_right_password_and_tou_acceptance(self) -> None:
         assert "eduPersonPrincipalName" in attributes
         assert attributes["eduPersonPrincipalName"] == [f"hubba-bubba@{self.app.conf.default_eppn_scope}"]
 
-    def test_login_missing_attributes(self) -> None:
+    def test_login_mfaauth(self) -> None:
         # pre-accept ToU for this test
         self.add_test_user_tou()
 
-        # remove mail address from user to simulate missing attribute
-        self.test_user.mail_addresses = MailAddressList()
-        self.request_user_sync(self.test_user)
+        # add security key to user
+        self.add_test_user_security_key()
 
         # Patch the VCCSClient, so we do not need a vccs server
         with patch.object(VCCSClient, "authenticate") as mock_vccs:
             mock_vccs.return_value = True
             result = self._try_login()
 
-        assert result.visit_order == [IdPAction.PWAUTH, IdPAction.FINISHED]
+        assert result.visit_order == [
+            IdPAction.PWAUTH,
+            IdPAction.MFA,
+            IdPAction.FINISHED,
+        ], f"NOT expected visit order {result.visit_order}"
+        assert result.sso_cookie_val is not None
+        assert result.finished_result is not None
+        assert result.finished_result.payload["message"] == IdPMsg.finished.value
+        assert result.finished_result.payload["target"] == "https://sp.example.edu/saml2/acs/"
+        assert result.finished_result.payload["parameters"]["RelayState"] == self.relay_state
+
+        attributes = self.get_attributes(result)
+
+        assert "eduPersonPrincipalName" in attributes
+        assert attributes["eduPersonPrincipalName"] == [f"hubba-bubba@{self.app.conf.default_eppn_scope}"]
+
+    def test_login_no_mandatory_mfa(self) -> None:
+        # pre-accept ToU for this test
+        self.add_test_user_tou()
+
+        # add security key to user
+        self.add_test_user_security_key(always_use_security_key_user_preference=False)
+
+        # Patch the VCCSClient, so we do not need a vccs server
+        with patch.object(VCCSClient, "authenticate") as mock_vccs:
+            mock_vccs.return_value = True
+            result = self._try_login()
+
+        assert result.visit_order == [
+            IdPAction.PWAUTH,
+            IdPAction.FINISHED,
+        ], f"NOT expected visit order {result.visit_order}"
+        assert result.sso_cookie_val is not None
         assert result.finished_result is not None
-        assert len(result.finished_result.payload["missing_attributes"]) == 1
-        assert result.finished_result.payload["missing_attributes"][0]["friendly_name"] == "mailLocalAddress"
+        assert result.finished_result.payload["message"] == IdPMsg.finished.value
+        assert result.finished_result.payload["target"] == "https://sp.example.edu/saml2/acs/"
+        assert result.finished_result.payload["parameters"]["RelayState"] == self.relay_state
+
```

Comment:
Should this and previous assert checks in this file also use the _check_login_result() function?